Refactor mods to support more general data types

GiggleLiu opened the pull request:
Nice package, I did some changes to fit my using case. List of changes are,
* add a second type parameter to Mod, so that `UInt` can be used in computing, `GaussMod{N,T}` is now an alias of `Mod{N,Complex{T}}`, so that a lot code can be reused.
* improve the performance of `Mod` algebra a lot.
* Clean up codes
    * use promotion instead of redefining algebras with integers and rationals.
* make CRT compatible with big integers, so that one does not suffer from integer overflow when using CRT to compute extremely big numbers. One the otherside, it will not return the modulus anymore, because the modulus can sometimes be a BigInt, which is not a valid type parameter.
* update documentations and README.

Looking forward to your review.

GiggleLiu committed:
save

GiggleLiu committed:
update 1

GiggleLiu committed:
tests pass

GiggleLiu committed:
big integer crt

GiggleLiu committed:
fix the algebra for unsigned

GiggleLiu commented:
I notice there is a better way to implement this. Let me polish it a bit before you can review  :P

GiggleLiu committed:
make GaussMod an alias of Mod

GiggleLiu commented:
Now I think this PR is ready for review @scheinerman . I would suggest reading the tests so that you can get what I have changed quickly. Hope these changes does not break your projects built on top of it.

GiggleLiu committed:
clean up functions

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
-    return mx
-end
+==(x::Mod{N,T1}, y::Mod{M,T2}) where {M,N,T1,T2} = false
+==(x::Mod{N,T1}, y::Mod{N,T2}) where {N,T1,T2} = value(x)==value(y)
```
This calls `mod(_, N)` twice. Wouldn't be like `iszero(x - y)` more efficient?

GiggleLiu replied:
You are right, I should define iszero instead of defining `==`.

KlausC commented on src/Mods.jl:
```diff
@@ -1,33 +1,33 @@
 module Mods
 
-import Base: isequal, (==), (+), (-), (*), (inv), (/), (//), (^), hash, show
-import Base: zero, one, rand, conj
```
Where are methods for `zero`, `one` defined now?
Are `iszero` and `oneunit` still missing?

GiggleLiu replied:
They are automatically generated. Since Mod subtypes Number, it will use `Mod{N,T}(0)` as the zero element and `Mod{N,T}(1)` as the one and oneunit. `iszero` can be defined by `== 0`. Since iszero is more efficient, I switch to defining `iszero` first and use `iszero` for defining `==` in the new commits.

KlausC replied:
ok.

scheinerman commented:
@GiggleLiu You did an amazing amount of work on this. Thank you.

scheinerman commented:
@GiggleLiu When I run the checks the documentation build doesn't work. I've never built documentation (I just use the `README` file) so I don't know what to do. Would you like to see what the problem is before I merge and release?

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
-function -(x::Mod{M}) where M
-    return Mod(-x.val, M)
+function -(x::Mod{M,T}) where {M,T<:Signed}
+    return Mod{M,T}(-x.val)  # Note: might break for UInt
```
that is broken, if `x.val === typemin(T)`.

GiggleLiu replied:
Good catch, fixed with a test case.

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
+    Mod{M,T}(_invmod(x.val, M))
+end
+_invmod(x::Unsigned, m::Unsigned) = invmod(x, m)
+@inline function _invmod(x::Signed, m::Signed)
```
Why can't we use `invmod` in `Signed` case like in `Unsigned` case ?

GiggleLiu replied:
Good point, `gcdx` is faster, but not applicable for `UInt`. Using `invmod` in both cases will cause regression in the performance of `Int` type.

KlausC replied:
I see. `invmod` has to check special cases wit negative arguments and performs a `mod`, which are not necessary in this use case. Maybe write a comment to avoid this question in the future,

GiggleLiu replied:
Sounds good, added some comments.

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
 
 """
 `inv(x::Mod)` gives the multiplicative inverse of `x`.
 This may be abbreviated by `x'`.
```
Where is this implemented. (`x'` compiles to `adjoint(x)`, which is not `inv(x)` ).

GiggleLiu replied:
Right, I removed this comment. For numbers, `adjoint `is same as `conj`. I see `conj` already defined in the file.

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
+function CRT(remainders, primes)
+    length(remainders) == length(primes) || error("size mismatch")
+    isempty(remainders) && return 1
+    M = prod(primes)
```
unhandled integer overflow, if type of `primes` is not `BigInt`. In the typical use case `primes` are small, but result is big.
Not sure if this algorithm is efficient, because all arithmetic need to be done in maximal size integers (of type `BigInt`) to be correct.

GiggleLiu replied:
This function does not change the element types automatically. If one wants to compute with BigInt, he should input BigInt vectors. Users are supposed to use other interfaces bellow.

KlausC replied:
Yes. But then a check for overflow of `M` is needed in order to not obtain false results without warning.

GiggleLiu replied:
This makes sense. I changed the default type to BigInt.
If a user want to use Int64 for computing, he should write `CRT(Int64, args...)`
There will be no overflow error, because the user is aware of issue by writting this explicitly.

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
+    isempty(rs) && return 1
+    CRT(convert.(T, value.(rs)), convert.(T, modulus.(rs)))
 end
+CRT(rs::Mod{<:Any, T}...) where T = CRT(T, rs...)
```
see above: result type maybe frequently too small. Better user `BigInt` as result type.

Coresponding test cases are missing (for the overflowing case).

GiggleLiu replied:
Do you mean making BigInt as the default?

KlausC replied:
That was my idea, but otherwise see above (check overflow).

KlausC commented on src/Mods.jl:
```diff
@@ -49,224 +49,130 @@ julia> value(a)
-    end
+function CRT(remainders, primes)
+    length(remainders) == length(primes) || error("size mismatch")
+    isempty(remainders) && return 1
```
Type stability issue: The returned `1` has not always the same type as the final return value (`typeof(M)`).

GiggleLiu replied:
I throwed an error instead. It does not make sense to feed an empty vector into this function. Good?

KlausC commented on test/runtests.jl:
```diff
@@ -38,12 +38,47 @@ end
+    @test rand(GaussMod{6}) isa GaussMod
+    @test eltype(rand(GaussMod{6}, 4, 4)) <: GaussMod
     p = 23
-    a = GaussMod{p}(3 - im)
```
Why do you ban `GaussMod` from the tests, it is stell defined and to be tested.

GiggleLiu replied:
I added this `GaussMod` back to make sure this PR is compatible with previous interface.

KlausC replied:
It is required in the tests because it is defined in the src.

GiggleLiu replied:
Yeah, I meant I added GaussMod back to the tests.

KlausC commented on src/GaussMods.jl:
```diff
@@ -5,74 +5,34 @@ export GaussMod, AbstractMod
+const GaussMod{N,T} = Mod{N,Complex{T}}
+GaussMod{N}(x::T) where {N, T<:Integer} = Mod{N,Complex{T}}(x)
+GaussMod{N}(x::T) where {N, T<:Complex} = Mod{N,T}(x)
+Mod{N}(x::Complex{Rational{T}}) where {N,T} = Mod{N}(real(x)) + Mod{N}(imag(x))*im
```
missing test case.

GiggleLiu replied:
Added the test. Wondering where did you see the test coverage? Can I change the travis CI to github actions and add the code coverage in this PR so that I can check the coverage more carefully?

TYI: Travis CI is becoming very restrictive on the amount of jobs one can submit, so people are switching to Github actions.

KlausC replied:
I did not "see" the missing test coverage, but did not find a test case in the test files, and I did not find a link to test coverage results in the README.

If you want to switch to github - CI, I give you a link to my current package; see directores `.ci`, `.github`, `.codecov.yml` and `README.md`. Most of the files in there have to be customized - but that is easy.

Probably the github actions have to be activated by the repository owner.

GiggleLiu replied:
Yeah, let's fix the CI in a seperate PR. Added the missing test case.

KlausC replied:
BTW: If you run the tests locally, ther is a coverage option:
```
]activate Mods
test --coverage
```
which leaves `*.jl.cov` files in `src`.

KlausC reviewed (changes requested):
Good improvements.
I would generally add a test coverage report, because it is not clear, if all branches of the code are proved by test cases.

scheinerman reviewed (approved):
I am a GitHub noob, so I'm not sure what I need to do. Is this ready for me to "merge" and how do I do that? Has the documentation error been fixed?

KlausC commented:
@scheinerman if you think, the changes are all ok, you can press the "merge" button; that will insert all changes into the main development branch.

Neverthless there are quite some undiscussed review comments, which should be resolved by the author of the PR, before you approve and merge the changes. You can add your comments to the individual review threads.

GiggleLiu committed:
resolve issues

GiggleLiu committed:
fix docs

GiggleLiu committed:
use iszero(x-y) instead of mod twice for comparison

GiggleLiu committed:
revert the previous change

GiggleLiu committed:
switch to defining iszero

GiggleLiu commented:
Hi, @scheinerman . I have fixed the documentation, the error was caused by the removal of docstrings of `GaussMod`, now I have added it back. Cheers.

And here is a not so important tip: After @KlausC approves the changes, you can merge the this PR. There is a dropdown menu along side the merge button, among different option to merge a PR in the dropdown menu, the "Squash and merge" option is recommended because it will keep the commit history clean.

scheinerman reviewed (commented):
@GiggleLiu Is this ready to be merged?

GiggleLiu commented:
@scheinerman I am still waiting for the feedback of @KlausC . We will ping you after all issues being resolved.

scheinerman commented:
> @scheinerman I am still waiting for the feedback of @KlausC . We will ping you after all issues being resolved.

Thanks @GiggleLiu . (This is all so confusing to me!)

KlausC commented on src/Mods.jl:
```diff
@@ -158,17 +160,16 @@ julia> 92%14
+function CRT(remainders, primes) where T
     length(remainders) == length(primes) || error("size mismatch")
-    isempty(remainders) && return 1
+    isempty(remainders) && throw(ArgumentError("input arguments should not be empty."))
```
A sensible return value would be `zero(eltype(primes))`.

GiggleLiu replied:
It is not intuitive to me whether it should return zero or one for empty inputs. Do you have any arguments favoring zero?

GiggleLiu replied:
BTW, there is another reason why I choose to throw an error. The `eltype` for a tuple of length 0 is ill-defined, while this interfaces aim to support both vectors and tuples.

KlausC replied:
If you specify the result type as the first argument (which defaults to `BigInt`) as you suppose, the type of the return value is clear.
Arguments for `0` as result for empty inputs are for example:

1. the condition `0 <= crt(r, p) < prod(p)` for empty `p` and `prod([]) == 1`
2. for integer arrays, the rest of the algorithm executes and delivers `0`.
3. it is a corner case in the identity `crt([r1; r2], [p1;p2]) == crt(crt(r1, p1), crt(r2, p2)], [lcm(p1); lcm(p2)])`

GiggleLiu replied:
Thanks for detailed explaination, it makes sense. Return 0 instead in the new commit.

KlausC reviewed (commented):
Still a few open issues.
Switching CI and test coverage is maybe worth a separate PR.

GiggleLiu committed:
resolve issues

KlausC commented on src/Mods.jl:
```diff
@@ -159,6 +162,12 @@ julia> 92%14
+    If you are confident that numbers does not overflow in your application,
+    please specify an optional type parameter as the first argument.
 """
 function CRT(remainders, primes) where T
```
There is no default `BigInt` argument, opposed to the note.

GiggleLiu replied:
Added the type argument in the new commit.

GiggleLiu committed:
fix CRT

KlausC reviewed (approved):
That was a good job! I am happy with the changes now.

scheinerman commented:
Dear @GiggleLiu  and @KlausC : I appreciate your interest and work on this `Mods` module. Is it ready to be merged??? (I'll be offline all day today -- Sunday -- so I will see your answers tomorrow.)

GiggleLiu commented:
@scheinerman Yes, it is ready to be merged. Please go ahead.
BTW, may I ask for a new release like 1.3.0 after merging this pr?

(No worries, have a nice weekend!)

scheinerman commented:
> @scheinerman Yes, it is ready to be merged. Please go ahead.
> BTW, may I ask for a new release like 1.3.0 after merging this pr?
> 
> (No worries, have a nice weekend!)

Yes. I'll test it out and then publish a release through JuliaHub at 1.3.0 as you suggested. There are enough changes that we should increase the middle number. Thank you again for your contributions!

scheinerman commented:
@GiggleLiu @KlausC @charleskawczynski  : I see that the `README` says that docs are passing, but where do I find the docs? (I've always just used the `README` to document modules.)

GiggleLiu commented:
> I see that the README says that docs are passing, but where do I find the docs?

I think there are some issue related to deployment, has the documentation been successfully deployed before? I can not find any existing version.

The current error message is 

```
 ! [remote rejected] HEAD -> gh-pages (cannot lock ref 'refs/heads/gh-pages': is at f96270b502aeea926c927aa672e5defe9506a77a but expected 7abb1c83c1bde0c0f81a83b07abbcf30f7e04028)
106
error: failed to push some refs to 'https://github.com/scheinerman/Mods.jl.git'
107
┌ Error: Failed to push:
108
│   exception =
109
│    failed process: Process(`git push -fq upstream HEAD:gh-pages`, ProcessExited(1)) [1]
110
│    
111
│    Stacktrace:
112
│      [1] pipeline_error
```
It is a git push command error, maybe a permission issue? I am not sure.


> (I've always just used the README to document modules.)

Me too. Actually this is a very good practice to host documents in README as long as a single page can do the job. Without documentation, one can still check the APIs in an REPL. So, no worry if the documentation issue can not be fixed.

scheinerman commented:
> > I see that the README says that docs are passing, but where do I find the docs?
> 
> I think there are some issue related to deployment, has the documentation been successfully deployed before? I can not find any existing version.
> 
> The current error message is
> 
> ```
>  ! [remote rejected] HEAD -> gh-pages (cannot lock ref 'refs/heads/gh-pages': is at f96270b502aeea926c927aa672e5defe9506a77a but expected 7abb1c83c1bde0c0f81a83b07abbcf30f7e04028)
> 106
> error: failed to push some refs to 'https://github.com/scheinerman/Mods.jl.git'
> 107
> ┌ Error: Failed to push:
> 108
> │   exception =
> 109
> │    failed process: Process(`git push -fq upstream HEAD:gh-pages`, ProcessExited(1)) [1]
> 110
> │    
> 111
> │    Stacktrace:
> 112
> │      [1] pipeline_error
> ```
> 
> It is a git push command error, maybe a permission issue? I am not sure.
> 
> > (I've always just used the README to document modules.)
> 
> Me too. Actually this is a very good practice to host documents in README as long as a single page can do the job. Without documentation, one can still check the APIs in an REPL. So, no worry if the documentation issue can not be fixed.

I think @charleskawczynski started setting this up for me. I may remove the line from the README that gives badges for documentation. @charleskawczynski : Do you want to set this up? Ok by me if you simply want to let it all slide.

charleskawczynski commented:
> has the documentation been successfully deployed before?

> I may remove the line from the README that gives badges for documentation. @charleskawczynski : Do you want to set this up?

No, it hasn't. There is still an issue with deployment. I'm able to merge PRs, but I don't have access to the settings, so I can't really help configure the deployment keys.

> Actually this is a very good practice to host documents in README as long as a single page can do the job. Without documentation, one can still check the APIs in an REPL. So, no worry if the documentation issue can not be fixed.

I think that's a good point, I was on the fence with suggesting to add the doc pages to start with, I think the docs are relatively short and just a readme is fine.
